### ISSUE TITLE
Selecting left while in visual multi cursor mode causes jumping cursors

### ISSUE THREAD

**Issue by jsonMartin** [VSCodeVim/Vim#1860]
* Click *thumbs-up* 👍 on this issue if you want it!
* Click *confused* 😕 on this issue if not having it makes VSCodeVim unusable. 

The VSCodeVim team prioritizes issues based on reaction count. 

--------

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug Report

**Environment**: MacOS

- **VSCode Version**: 1.13.1
- **VsCodeVim Version**: 0.8.6
- **OS**: 10.12.5

**What happened**: Selecting left while in visual multi cursor mode causes jumping cursors

**What did you expect to happen**: Multiple cursors in visual mode should be able to select left as well as right

**How to reproduce it**: Add multiple cursors in visual mode, and try to select left

![large gif 640x377](https://user-images.githubusercontent.com/4033621/27412913-01f544dc-56ad-11e7-868c-83cf6f83ba9e.gif)

**Comment by westim** [comment]
Hello! New contributor here :)

I believe I may have a fix for this issue. Mind if I take it on?

**Comment by Chillee** [comment]
@westim Of course!

**Comment by westim** [comment]
In the first commit on my fork, I simply changed the `Position.FromVSCodeSelection()` method to use the `Selection.anchor` and `Selection.active` properties instead of the `Selection.start` and `Selection.end` properties, respectively. This appears to have eliminated the cursor jitter we see above.

![Demo](https://thumbs.gfycat.com/CircularAcclaimedIndigobunting-size_restricted.gif)

As you can see in the GIF above, this change has introduced a new problem: the Selection gets extend to the EOL to the right. I'm not too familiar with the workflow of this extension, but it appears that the action isn't considered "done" after the first move as expected and instead keeps calling the `updateView()` method until the `cursorStart` position is bumped all the way to the EOL by the following code in modeHandler.ts:

```Typescript
      } else {
        // MultiCursor mode is active.

        if (vimState.currentMode === ModeName.Visual) {
          selections = [];

          for (let { start: cursorStart, stop: cursorStop } of vimState.allCursors) {
            if (cursorStart.compareTo(cursorStop) > 0) {
              cursorStart = cursorStart.getRight();

            selections.push(new vscode.Selection(cursorStart, cursorStop));
          }
```
Removing the `if` statement comparing the cursor positions results in the following behavior:

![Second Attempt](https://thumbs.gfycat.com/ParchedTornFlickertailsquirrel-size_restricted.gif)

This is *almost* the behavior we want, except the selection is offset to the left by one when selecting to the left. Presumably, this is why that `if` statement existed in the first place.

I'm a bit worried that I'm starting to dig into delicate portions of the codebase, but none of the existing unit tests appear to be failing due to these changes. I'll look into adding tests once I find a sufficient solution.

**Comment by westim** [comment]
I've spent a good amount of time on this and believe I've arrived at a sufficient solution.

# Problem

Ultimately, the problem was caused by this bit of code in `handleSelectionChange()`:
```Typescript
// Number of selections changed, make sure we know about all of them still
this.vimState.allCursors = e.textEditor.selections.map(
x => new Range(Position.FromVSCodePosition(x.start), Position.FromVSCodePosition(x.end))
```
This code is problematic because the assumption that you can map directly from `Selections[]` to `vimState.AllCursors` is incorrect. This is evident by the extra work performed by `updateView()`:
```Typescript
        // MultiCursor mode is active.

        if (vimState.currentMode === ModeName.Visual) {
          selections = [];

          for (let { start: cursorStart, stop: cursorStop } of vimState.allCursors) {
            if (cursorStart.compareTo(cursorStop) > 0) {
              cursorStart = cursorStart.getRight();

            selections.push(new vscode.Selection(cursorStart, cursorStop));
          }
```
This code snippet does the opposite operation, converting from `vimState.allCursors` to `Selections[]`, while making adjustments to the cursor positions when the cursor is going to be moved to the left.
# Solution
The solution is fairly evident: simply perform the opposite adjustments in `handleSelectionChange()`:
```Typescript
      // Number of selections changed, make sure we know about all of them still
      let cursors: Range[];
      cursors = [];
      for (let sel of e.textEditor.selections) {
        let Start = Position.FromVSCodePosition(sel.anchor);
        let Stop = Position.FromVSCodePosition(sel.active);
        if (Start.compareTo(Stop) > 0) {
          Start = Start.getLeft();
        }
        cursors.push(new Range(Start, Stop));
      }
      this.vimState.allCursors = cursors;
```
This results in the expected behavior.
# Testing
Manual testing was performed to confirm that this issue has been resolved. I also confirmed that all unit tests pass (although I don't believe the extension has any multi-cursor unit tests).

I also performed the manual testing suggested in the header comment for `handleSelectionChange()`, which recommended running [these](https://gist.github.com/rebornix/d21d1cc060c009d4430d3904030bd4c1) manual tests to avoid regressions. No regressions were identified.

I went one step further and looked for other potential regressions using git blame. #1312 appears to be a regression, although not caused by these changes. #1432 is still fixed.

**Comment by xconverge** [comment]
Fixed by #2036
